best-practices/go-ci.md

```diff
@@ -0,0 +1,54 @@
+## Best practices for Continuous Integration with Go
+
+First, use [Modules](https://golang.org/ref/mod) even if you don't have any
+external dependencies. Beyond ensuring consistency, a module allows using new Go
+language features. The instructions below assume you are setting up CI for a
+single Go module.
+
+It is strongly recommended that the following checks are used:
+
+* `go test ./...` to run all tests
+* `go test -race ./...` to catch any data races when running the tests
+* `test -z $(gofmt -l .)` to ensure all Go files are well formatted
+* `go vet ./...` to ensure there are no vet warnings
+
+The following checks are also recommended, though they might not help in all
+situations:
+
+* `go mod tidy && git diff --exit-code` to ensure `go.mod` and `go.sum` are clean
+* `go mod verify` to ensure that `go.sum` agrees with what's in the module cache
+* `go test -run=- -bench=. -benchtime=1x ./...` to catch stale benchmarks
+* `staticcheck ./...` to ensure there are no [Staticcheck](https://staticcheck.io/) warnings
+
+Finally, if your module is a library, it is recommended to support at least the
+two latest major Go versions. For example, if Go 1.16.x is the latest major
+release, all the checks above should be run on 1.16.x, and `go test ./...`
+should also run on 1.15.x. Note that upstream Go supports
+[three major releases](https://github.com/golang/go/wiki/Go-Release-Cycle#release-maintenance)
+at once.
+
+#### Open questions and TODOs
+
+Should we recommend using `go fmt ./...` instead of `gofmt .`? The main
+difference is that the former will ignore any files not part of `./...`, such as
+Go files behind `testdata/` or `_ignored`.
+
+Should we mention popular static analysis frameworks like golangci-lint, since
+it will be a common question? We likely don't want to encourage their use, given
+their huge complexity. Plus, staticcheck gives most of the benefit with fewer
+false positives and better defaults.
+
+A good place to draw a line regarding external tools might be what gopls
+natively supports. Right now, that is just staticcheck and gofumpt.
+
+We should provide ready-to-copy-paste scripts with the above. We likely don't
+want to maintain examples in many popular CI config flavors, though. Perhaps
+link to external guides with more platform-specific details once they add a full
+script, like https://github.com/mvdan/github-actions-golang.
```

Comment:
In the context of Protocol Labs, you're right that this paragraph does not really apply. This paragraph is here in the context of this document being generic to any and all Go programmers, since the intent is to eventually upstream it to the official Go docs (see the link available elsewhere).

We could and still should link to your work, though, for PL readers.

Reply:
Done, in a new section.